// @zag-js/date-input/dist/utils/validity.mjs
import "../chunk-QZ7TP4HQ.mjs";

// src/utils/validity.ts
function getGroupCount(selectionMode) {
  return selectionMode === "range" ? 2 : 1;
}
function getGroupOffset(allSegments, index) {
  return allSegments.slice(0, index).reduce((acc, segs) => acc + segs.length, 0);
}
function resolveActiveSegment(ctx) {
  const { context, computed } = ctx;
  const index = context.get("activeIndex");
  const activeSegmentIndex = context.get("activeSegmentIndex");
  const allSegments = computed("segments");
  const offset = getGroupOffset(allSegments, index);
  return { allSegments, segments: allSegments[index], offset, localIndex: activeSegmentIndex - offset };
}
function getActiveSegment(ctx) {
  const { allSegments, localIndex } = resolveActiveSegment(ctx);
  const index = ctx.context.get("activeIndex");
  return allSegments[index]?.[localIndex];
}
function advanceToNextSegment(ctx) {
  const { context } = ctx;
  const index = context.get("activeIndex");
  const { allSegments, segments, offset, localIndex } = resolveActiveSegment(ctx);
  const nextLocalIndex = segments.findIndex((s, i) => i > localIndex && s.isEditable);
  if (nextLocalIndex !== -1) {
    context.set("activeSegmentIndex", offset + nextLocalIndex);
    return;
  }
  const nextGroupSegments = allSegments[index + 1];
  if (!nextGroupSegments) return;
  const firstNextGroupEditableLocalIndex = nextGroupSegments.findIndex((s) => s.isEditable);
  if (firstNextGroupEditableLocalIndex === -1) return;
  context.set("activeIndex", index + 1);
  context.set("activeSegmentIndex", offset + segments.length + firstNextGroupEditableLocalIndex);
}
function getActiveDisplayValue(ctx) {
  const index = ctx.context.get("activeIndex");
  return ctx.context.get("displayValues")[index];
}
function setDisplayValue(ctx, index, dv) {
  const displayValues = [...ctx.context.get("displayValues")];
  displayValues[index] = dv;
  ctx.context.set("displayValues", displayValues);
}
export {
  advanceToNextSegment,
  getActiveDisplayValue,
  getActiveSegment,
  getGroupCount,
  getGroupOffset,
  resolveActiveSegment,
  setDisplayValue
};
